### PR TITLE
extensible-effects conversion

### PR DESCRIPTION
**Pull request by theotherjimmy**
This was much harder to understand, and much easier to implement than I thought.

### CODE REVIEW

**Comment by theotherjimmy** [comment]
added the commit so that @fosskers and @bb010g can take a look.
But good god the compile errors right now.

File: src/Aura/Shell.hs [hunk 1]
```diff
@@ -68,12 +72,49 @@ isTrueRoot :: Sh Bool
 isTrueRoot = (&&) <$> (not <$> varExists "SUDO_USER") <*> u
   where u = (== Just "root") <$> get_env "USER"
 
+isntTrueRoot :: Sh Bool
+isntTrueRoot = not <$> isntTrueRoot
```
Comment (theotherjimmy) [inline src/Aura/Shell.hs]: I actually did that.

File: src/Aura/Commands/A.hs [hunk 2]
```diff
@@ -103,43 +105,43 @@ auraCheck toUpgrade = if "aura" `elem` toUpgrade
...
-    liftIO $ putStrLn $ renderAurPkgInfo ss ai ns <> "\n"
+    let name = aurNameOf ai
+    ns <- fromJust <$> pkgbuild' name >>= namespace name
+    liftIO $ IO.putStrLn $ renderAurPkgInfo ss ai ns <> "\n"
 
-renderAurPkgInfo :: Settings -> AurInfo -> Namespace -> String
+renderAurPkgInfo :: Settings -> AurInfo -> Namespace -> T.Text
 renderAurPkgInfo ss ai ns = entrify ss fields entries
     where fields   = fmap bForeground . infoFields . langOf $ ss
-          empty x  = case x of [] -> "None"; _ -> x
```
Comment (fosskers) [inline src/Aura/Commands/A.hs]: Cute :)

**Comment by fosskers** [comment]
It looks like this encompasses the `Text` conversion as well. I'll do some tweaks afterward to make the `Shelly` use a bit more seamless, and that should be much easier now.

**Comment by theotherjimmy** [comment]
Yep. I wanted to get it compiling to check that ext-eff conversion was working. I can confirm now that it is, but it took 1100 lines to check ~30.

so the three major changes in this PR:
String ==> Text
finished Shelly (no more Shell)
MTL ==> Extensible Effects.

File: src/Aura/Build.hs [hunk 3]
```diff
@@ -76,23 +80,24 @@ build' :: [Buildable] -> Aura ([FilePath], [Buildable])
 build' []     = failure "build' : You should never see this."
 build' (p:ps) = ask >>= \ss -> do
   let user = buildUserOf ss
-  curr   <- liftIO pwd
+      quiet = suppressMakepkg ss
+  curr   <- liftShelly pwd
   getBuildScripts p user curr
   overwritePkgbuild p
-  pNames <- join (makepkg <*> pure user)
+  pNames <- liftShelly (makepkg quiet user []) >>= wrap
```
Comment (theotherjimmy) [inline src/Aura/Build.hs]: Should not use the empty list here, instead actually pass arguments to makepkg. it's there for type checking purposes right now.

File: src/Aura/Commands/A.hs [hunk 4]
```diff
@@ -147,52 +149,57 @@ aurPkgSearch (fold -> regex) = ask >>= \ss -> do
...
+                                                   (Re.suffix 0 match)))
+                    <$> Re.find (Re.regex [] ("(?i)"<> r)) cs)
           repo = magenta "aur/"
-          n = c bForeground $ T.unpack $ aurNameOf i
-          d = c noColour $ T.unpack $ aurDescriptionOf i
-          l = yellow . show $ aurVotesOf i  -- `l` for likes?
-          p = yellow $ printf "%0.2f" (popularityOf i)
+          n = c bForeground $ aurNameOf i
+          d = c noColour $ aurDescriptionOf i
+          l = yellow . T.pack . show $ aurVotesOf i  -- `l` for likes?
```
Comment (theotherjimmy) [inline src/Aura/Commands/A.hs]: This should be something like sformat int

File: src/Aura/Commands/B.hs [hunk 5]
```diff
@@ -34,14 +35,12 @@ import Aura.Monad.Aura
...
-import Shell (rm)
-
 ---
 
 -- Pretty similar to `-Cc`...
-cleanStates :: [String] -> Aura ()
+cleanStates :: [T.Text] -> Aura ()
 cleanStates []        = cleanStates' 0
-cleanStates (input:_) | all isDigit input = cleanStates' $ read input
+cleanStates (input:_) | T.all isDigit input = cleanStates' $ read $ T.unpack input
```
Comment (theotherjimmy) [inline src/Aura/Commands/B.hs]: Should use Data.Text.Read like C.hs:128

File: src/Aura/Commands/L.hs [hunk 6]
```diff
@@ -40,44 +38,47 @@ import Aura.Utils       (entrify)
...
+          installDate =  Re.find (Re.regex [] "\\[[-:0-9 ]+\\]") (head matches) >>= Re.group 0
           upgrades    = length $ searchLines " upgraded " matches
-          recent      = unlines . fmap (" " <>) . takeLast 5 $ matches
+          recent      = T.unlines . fmap (" " <>) . takeLast 5 $ matches
           takeLast n  = reverse . take n . reverse
           entries     = [ pkg
-                        , installDate
-                        , show upgrades
+                        , fromMaybe "" installDate
+                        , T.pack $ show upgrades
```
Comment (theotherjimmy) [inline src/Aura/Commands/L.hs]: Should use Formatting

File: src/Aura/Commands/O.hs [hunk 7]
```diff
@@ -34,9 +36,9 @@ import Aura.Monad.Aura
 
 ---
 
-displayOrphans :: [String] -> Aura ()
-displayOrphans []   = orphans >>= liftIO . traverse_ putStrLn
+displayOrphans :: [T.Text] -> Aura ()
+displayOrphans []   = orphans >>= liftIO . traverse_ (putStrLn . show)
```
Comment (theotherjimmy) [inline src/Aura/Commands/O.hs]: should use Data.Text.IO and IO.putStrLn instead of show

File: src/Aura/Conflicts.hs [hunk 8]
```diff
@@ -54,16 +57,16 @@ realPkgConflicts ss pkg dep
           lang     = langOf ss
           toIgnore = ignoredPkgsOf ss
           failMsg1 = getRealPkgConflicts_2 name lang
-          failMsg2 = getRealPkgConflicts_1 name curVer (show reqVer) lang
+          failMsg2 = getRealPkgConflicts_1 name curVer (T.pack $ show reqVer) lang
```
Comment (theotherjimmy) [inline src/Aura/Conflicts.hs]: Use Formatting

File: src/Aura/Core.hs [hunk 9]
```diff
@@ -36,59 +34,61 @@ import Aura.Languages
...
+                   | MustBe T.Text
                    | Anything
                      deriving (Eq)
 
 instance Show VersionDemand where
-    show (LessThan v) = "<" <> v
-    show (AtLeast v)  = ">=" <> v
-    show (MoreThan v) = ">" <> v
-    show (MustBe  v)  = "=" <> v
+    show (LessThan v) = show ("<" <> v)
```
Comment (theotherjimmy) [inline src/Aura/Core.hs]: Formatting?

File: src/Aura/Core.hs [hunk 10]
```diff
@@ -124,77 +123,78 @@ parseDep s = Dep name (getVersionDemand comp ver)
...
     then action else scoldAndFail trueRoot_3
 
 -- `-Qm` yields a list of sorted values.
 -- | A list of non-prebuilt packages installed on the system.
-foreignPackages :: Aura [(String, String)]
-foreignPackages = (fmap fixName . lines) <$> pacmanOutput ["-Qm"]
-    where fixName = hardBreak (' ' ==)
+foreignPackages :: Aura [(T.Text, T.Text)]
+foreignPackages = (fmap fixName . T.lines) <$> pacmanOutput ["-Qm"]
+    where fixName = (\ (name, ver) -> (name, T.stripStart ver)) . T.breakOn " "
```
Comment (theotherjimmy) [inline src/Aura/Core.hs]: T.breakon " " is T.unwords

File: src/Aura/Core.hs [hunk 11]
```diff
@@ -124,77 +123,78 @@ parseDep s = Dep name (getVersionDemand comp ver)
...
-removePkgs :: [String] -> [String] -> Aura ()
+removePkgs :: [T.Text] -> [T.Text] -> Aura ()
 removePkgs [] _         = pure ()
 removePkgs pkgs pacOpts = pacman  $ ["-Rsu"] <> pkgs <> pacOpts
 
 -- Moving to a libalpm backend will make this less hacked.
 -- | True if a dependency is satisfied by an installed package.
 isSatisfied :: Dep -> Aura Bool
-isSatisfied (Dep name ver) = null <$> pacmanOutput ["-T", name <> show ver]
+isSatisfied (Dep name ver) = T.null <$> pacmanOutput ["-T", name <> T.pack (show ver)]
```
Comment (theotherjimmy) [inline src/Aura/Core.hs]: Formatting

File: src/Aura/Languages.hs [hunk 12]
```diff
@@ -1036,8 +1039,8 @@ backupCache_4 (bt -> dir) = \case
     Indonesia  -> "Melakukan `backup` pada direktori " <> dir
     _          -> "Backing up cache to " <> dir
 
-backupCache_5 :: Int -> Language -> String
-backupCache_5 (bt . show -> n) = \case
+backupCache_5 :: Int -> Language -> T.Text
+backupCache_5 (bt . T.pack . show -> n) = \case
```
Comment (theotherjimmy) [inline src/Aura/Languages.hs]: Formatting

File: src/Aura/Languages.hs [hunk 13]
```diff
@@ -1104,8 +1107,8 @@ backupCache_8 = \case
     Indonesia  -> "Melakukan `backup`. Proses ini akan berjalan untuk beberapa menit..."
     _          -> "Backing up. This may take a few minutes..."
 
-copyAndNotify_1 :: Int -> Language -> String
-copyAndNotify_1 (cyan . show -> n) = \case
+copyAndNotify_1 :: Int -> Language -> T.Text
+copyAndNotify_1 (cyan . T.pack . show -> n) = \case
```
Comment (theotherjimmy) [inline src/Aura/Languages.hs]: Formatting

File: src/Aura/Languages.hs [hunk 14]
```diff
@@ -1172,8 +1175,8 @@ cleanCache_2 = \case
     Indonesia  -> "Akan menghapus SEMUA `cache` paket"
     _          -> "This will delete the ENTIRE package cache."
 
-cleanCache_3 :: Int -> Language -> String
-cleanCache_3 (bt . show -> n) = \case
+cleanCache_3 :: Int -> Language -> T.Text
+cleanCache_3 (bt . T.pack . show -> n) = \case
```
Comment (theotherjimmy) [inline src/Aura/Languages.hs]: Formatting

File: src/Aura/Languages.hs [hunk 15]
```diff
@@ -1255,8 +1258,8 @@ cleanNotSaved_1 = \case
     _          -> "Determining unneeded package files..."
 
 -- NEEDS TRANSLATION
-cleanNotSaved_2 :: Int -> Language -> String
-cleanNotSaved_2 (cyan . show -> s) = \case
+cleanNotSaved_2 :: Int -> Language -> T.Text
+cleanNotSaved_2 (cyan . T.pack . show -> s) = \case
```
Comment (theotherjimmy) [inline src/Aura/Languages.hs]: Formatting

File: src/Aura/Packages/ABS.hs [hunk 16]
```diff
@@ -92,59 +95,59 @@ makeBuildable repo name = do
                  , isExplicit   = False
                  , buildScripts = \fp -> Just <$> copyTo repo name fp }
 
-copyTo :: String -> String -> FilePath -> IO FilePath
+copyTo :: T.Text -> T.Text -> FilePath -> Aura FilePath
 copyTo repo name fp = do
-    void $ Sh.quietShellCmd "cp" ["-R", loc, fp]
-    pure $ fp </> name
-        where loc = absBasePath </> repo </> name
+    _ <- A.quietShellCmd "cp" ["-R",toTextIgnore loc, toTextIgnore fp]
```
Comment (theotherjimmy) [inline src/Aura/Packages/ABS.hs]: This seems to be bad style, just get rid of _ <-

File: src/Aura/Packages/ABS.hs [hunk 17]
```diff
@@ -92,59 +95,59 @@ makeBuildable repo name = do
...
 
 absBasePath :: FilePath
 absBasePath = "/var/abs"
 
 -- | All repos with all their packages in the local tree.
 absTree :: Aura ABSTree
-absTree = liftIO $ do
-    repos <- ls'' absBasePath >>= filterM doesDirectoryExist
+absTree = liftShelly $ do
+    repos <- ls'' absBasePath >>= filterM test_d
```
Comment (theotherjimmy) [inline src/Aura/Packages/ABS.hs]: this will fail when/if an empty string is passed to test_d

File: src/Aura/State.hs [hunk 18]
```diff
@@ -96,21 +100,21 @@ toChangeAndRemove old curr = M.foldrWithKey status ([], []) $ pkgsOf curr
...
+getStateFiles :: Aura [T.Text]
+getStateFiles = sort <$> liftShelly (lsT' stateCache)
 
 saveState :: Aura ()
 saveState = do
   state <- currentState
-  let filename = stateCache </> dotFormat (timeOf state)
-  liftIO $ writeFile filename (show state)
+  let filename = stateCache </> fromText (T.pack (dotFormat (timeOf state)))
+  liftIO $ writeFile filename $ E.encodeUtf8 $ T.pack (show state)
```
Comment (theotherjimmy) [inline src/Aura/State.hs]: Formatting

File: src/Aura/State.hs [hunk 19]
```diff
@@ -119,17 +123,19 @@ restoreState = ask >>= \ss -> do
   reinstallAndRemove (getFilename cache `mapMaybe` okay) remo
 
 readState :: FilePath -> Aura PkgState
-readState name = liftIO (read <$> readFileUTF8 (stateCache </> name))
+readState name = liftIO ((read . T.unpack) <$> readFileUTF8 (stateCache </> name))
```
Comment (theotherjimmy) [inline src/Aura/State.hs]: Should use Data.Text.Read like C.hs:128

File: src/Aura/Commands/L.hs (hunk 6 above)
Comment (fosskers) [inline src/Aura/Commands/L.hs]: Should use `basic-prelude`, where `show :: Show t => t => Text`, but that's for another PR.

**Comment by fosskers** [comment]
This is wonderful. Do you consider this good to go?

**Comment by theotherjimmy** [comment]
I do now. :D

**Comment by fosskers** [comment]
Do `extensible-effects` and `type-aligned` not have to be added to `extra-deps` in the `stack.yaml`? Travis seems to be building it fine regardless, though.

**Comment by theotherjimmy** [comment]
I don't know. I only use cabal.

**Comment by fosskers** [comment]
Ah, I see. Then that's something I can fix on this end right away. `stack` is amazing, btw, as @bb010g kept telling me.

**Pull request merged**